src/lib.rs

```diff
@@ -1,36 +1,75 @@
-use std::ptr::null_mut;
-use std::os::raw::c_int;
+mod libxenevtchn;
+
+extern crate nix;
+extern crate xenevtchn_sys;
+use libxenevtchn::LibXenEvtchn;
+use nix::poll::PollFd;
+use nix::poll::PollFlags;
 use std::io::Error;
+use std::os::raw::c_int;
+use std::ptr::null_mut;
+use xenevtchn_sys::{evtchn_port_t, xenevtchn_handle, xenevtchn_port_or_error_t};
 
 #[derive(Debug)]
 pub struct XenEventChannel {
-    handle: *mut xenevtchn_sys::xenevtchn_handle,
-    fd: c_int,
-    bind_port: c_int,
+    pub handle: *mut xenevtchn_sys::xenevtchn_handle,
+    pub fd_struct: PollFd,
+    pub bind_port: c_int,
+    libxenevtchn: LibXenEvtchn,
 }
 
 impl XenEventChannel {
+    pub fn new(domid: u32, evtchn_port: u32) -> Result<Self, Error> {
+        let handle = unsafe { xenevtchn_sys::xenevtchn_open(null_mut(), 0) };
+        if handle == null_mut() {
+            return Err(Error::last_os_error());
+        }
+        let fd = unsafe { xenevtchn_sys::xenevtchn_fd(handle) };
```

Comment:
move getting the file descriptor and preparing the poll struct in the Xen driver::listen() function.

Reply:
basically, polling should be not be the responsability of this crate.
also, it shouldn't depend on the nix crate